langkit/ocaml_api.py

```diff
@@ -0,0 +1,444 @@
+        """
+        Return a string representing the polymorphic variant for the given
+        concrete type.
+
+        :param ct.ASTNodeType type: The type for which we want the
+            polymorphic variant name.
+        :rtype: list[ct.CompiledType]
+        """
+        return "`{}".format(type.entity.api_name.camel)
+
+    def fields_name(self, for_module, from_module=None):
+        """
+        Return a string representing the name of the ocaml type for the
+        fields of a node.
+
+        :param ct.ASTNodeType for_module: The field comes from this module.
+        :param ct.ASTNodeType from_module: The field is accessed inside this
+            module.
+        :rtype: list[ct.CompiledType]
+        """
+        if for_module == from_module:
+            return 'fields'
+        else:
+            return "{}.{}".format(self.module_name(for_module), 'fields')
+
+    def check_for_null(self, c_value, astnode, context):
+        """
+        Return an ocaml expression that first check if the underlying node of
+        c_value is null. If it is the case, the expression is evaluated to
+        None, otherwise, wrap c_value to the given astnode type.
+
+        :param str c_value: The ocaml c value for a node.
+        :param ct.ASTNodeType: The type of c_value.
+        :param str context: The ocaml expression denoting the context.
+        """
+
+        return (
+            "if is_null (getf {} EntityStruct.node) then None else Some ({})".
+            format(c_value, self.wrap_value(c_value, astnode, context))
+        )
+
+    def get_field_type(self, field):
+        """
+        Return a precice type for the given Field instance. The list contains
+        the possible concrete types of the given field. If there is only
+        one element, the only element can be abstract, this means that the
+        type is precise enough.
+
+        :param ct.Field field: The field for which we want a precice type.
+        :rtype: list[ct.CompiledType]
+        """
+
+        precise_types = field.precise_types
+        concrete_precise_types = list(frozenset(
+            tpe
+            for precise_type in precise_types.matched_types
+            for tpe in precise_type.concrete_subclasses
+        ))
+
+        # If the list of concrete precise types has the same length as the list
+        # of concrete subclasses of the minimal common ancestor type, we can
+        # replace the list of concrete precise types by a singleton list
+        # containing the common ancestor type.
+        super_type = precise_types.minimal_common_type
+        if len(super_type.concrete_subclasses) == len(concrete_precise_types):
+            concrete_precise_types = [super_type]
+
+        return concrete_precise_types
+
+    def get_parse_fields(self, tpe):
+        """
+        Return all the parse fields to be exposed as a field in record of tpe.
+
+        :param ct.ASTNodeType tpe: The ast node type of which we want to get
```

Comment:
We haven’t renamed all the identifers yet, but for new names/docs we try to avoid mentioning “AST” since the tree that Langkit handles is a concrete syntax tree. So I suggest rewording to just “The node of which […]”. Also we generally don’t abbreviate `tpe` this way :-) I would suggest to rename it to `node` as it’s even more specific than any type.

Reply:
(Same comment for other occurences)